accounts/keystore/presale.go

```diff
@@ -0,0 +1,146 @@
+	"golang.org/x/crypto/pbkdf2"
+)
+
+func importPreSaleKey(keyStore keyStore,
+	keyJSON []byte, password string) (accounts.Account, encryptedKeyJSONV3, error) {
+	key, err := decryptPreSaleKey(keyJSON, password)
+	if err != nil {
+		return accounts.Account{}, encryptedKeyJSONV3{}, err
+	}
+
+	key.ID, err = uuid.NewRandom()
+	if err != nil {
+		return accounts.Account{}, encryptedKeyJSONV3{}, err
+	}
+
+	a := accounts.Account{
+		Address: key.Address,
+	}
+
+	encryptKey, err := keyStore.StoreKey(key, password)
+
+	return a, encryptKey, err
+}
+
+func decryptPreSaleKey(fileContent []byte, password string) (key *Key, err error) {
+	preSaleKeyStruct := struct {
+		EncSeed string
+		EthAddr string
+		Email   string
+		BtcAddr string
+	}{}
+
+	err = json.Unmarshal(fileContent, &preSaleKeyStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	encSeedBytes, err := hex.DecodeString(preSaleKeyStruct.EncSeed)
+	if err != nil {
+		return nil, errors.New("invalid hex in encSeed")
+	}
+
+	if len(encSeedBytes) < 16 {
+		return nil, errors.New("invalid encSeed, too short")
+	}
+
+	iv := encSeedBytes[:16]
+	cipherText := encSeedBytes[16:]
+	passBytes := []byte(password)
+	derivedKey := pbkdf2.Key(passBytes, passBytes, 2000, 16, sha256.New)
+	plainText, err := aesCBCDecrypt(derivedKey, cipherText, iv)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ethPriv := crypto.Keccak256(plainText)
+
+	ecKey, err := crypto.DToECDSA(ethPriv, false)
+	if err != nil {
+		return nil, err
+	}
+
+	key = &Key{
+		ID:         uuid.UUID{},
+		Address:    crypto.PubKeyToAddress(&ecKey.PublicKey),
+		PrivateKey: ecKey,
+	}
+
+	derivedAddr := hex.EncodeToString(key.Address.Bytes()) // needed because .Hex() gives leading "0x"
+	expectedAddr := preSaleKeyStruct.EthAddr
+
+	if derivedAddr != expectedAddr {
+		err = fmt.Errorf("decrypted addr '%s' not equal to expected addr '%s'", derivedAddr, expectedAddr)
+	}
+
+	return key, err
+}
+
+func aesCTRXOR(key, inText, iv []byte) ([]byte, error) {
+	// AES-128 is selected due to size of encryptKey.
+	aesBlock, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	stream := cipher.NewCTR(aesBlock, iv)
+	outText := make([]byte, len(inText))
+	stream.XORKeyStream(outText, inText)
+
+	return outText, err
+}
+
+func aesCBCDecrypt(key, cipherText, iv []byte) ([]byte, error) {
+	aesBlock, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	decrypter := cipher.NewCBCDecrypter(aesBlock, iv)
+	paddedPlaintext := make([]byte, len(cipherText))
+	decrypter.CryptBlocks(paddedPlaintext, cipherText)
+
+	plaintext := pkcs7Unpad(paddedPlaintext)
+	if plaintext == nil {
+		return nil, accounts.ErrDecrypt
+	}
+
+	return plaintext, err
+}
+
+// From https://leanpub.com/gocrypto/read#leanpub-auto-block-cipher-modes
+func pkcs7Unpad(in []byte) []byte {
+	if len(in) == 0 {
+		return nil
+	}
+
+	padding := in[len(in)-1]
+	if int(padding) > len(in) || padding > aes.BlockSize {
+		return nil
+	} else if padding == 0 {
+		return nil
+	}
```

Comment:
It seems else if is obsolete here?
```suggestion
	if int(padding) > len(in) || padding > aes.BlockSize || padding == 0 {
		return nil
	}
```